react/ComponentStyle.md

```diff
@@ -0,0 +1,304 @@
+# Stateless Functional Component
+
+* [when to use?](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/prefer-stateless-function.md)
+  * no lifecycle method
+  * no state to keep, only render data from props.
+
+* code snippet
+
+```
+// Madadata.jsx
+import React, { PropTypes } from 'react';
+
+// you can do this with function declaration. (function Madadata = ())
+const Madadata = props => <div>{props.name}</div>;
+
+Madadata.propTypes = {
+  name: PropTypes.string.isRequired,
+};
+
+// don't use context if you don't know what it is.
+// Madadata.contextTypes = {}
+
+export default Madadata;
+```
+
+* signature
+
+```
+(props, context) => JSXElement;
+```
+
+* Notes
+  * capitalize the **first character** of your component.
+  * optional params of this function can be **props**, **context**.
+  * specify the propTypes of your component for props and context **explicitly**.
+  * export your component as **default**.
+  * name the file as **ComponentName.jsx**.
+
+* Styles
+  * prefer arrow function
+  * If JSX spans for several lines
+    ```
+    const Madadata = props => (
+      <div>
+        {props.name}
+      </div>
+    );
+    ```
+  * prefer collapsing props in propTypes and contextTypes and prefer isRequired to enforce checking prop existence.
+    ```
+    Madadata.propTypes = {
+      name: PropTypes.string.isRequired,
+    };
+
+    Madadata.contextTypes = {
+      store: PropTypes.object.isRequired,
+    }
+    ```
+
+# PureComponent
+
+* when to use?
+  * has lifecycle method
+  * has state to keep.
+  * UI is determined only by state, props.
+    * UI do not depend on global variable.
+    * UI do not use any utility function that has variables from outside.
+    * UI do not have children that are not PureComponent.
+
+* code snippet
+
+```
+// PureMadadata.jsx
+import React, { PureComponent, PropTypes } from 'react';
+
+class PureMadadata extends PureComponent {
+
+  constructor(props, context) {
+    super(props, context);
+    this.state = { value: 'madadata' };
+    this.handleSomethingClick = this.handleSomethingClick.bind(this);
+  }
+
+  componentWillMount() {
+    // * call this.setState here to update state, the next render will pick up state changes.
```

Comment:
nice catch. I mean the next render is the first render. But obviously first render is more clear